### ISSUE TITLE
Update MPI I/O support + Requires

### ISSUE THREAD

**Issue by jipolanco** [JuliaIO/HDF5.jl#619]
The `test/mpio.jl` script fails with recent versions of MPI.jl.

While I started working on a fix, I was thinking that it would be nice to have a friendlier way of calling MPI-IO functions such as `h5p_set_fapl_mpio`. Right now, one has to pass low-level `MPI_Comm` and `MPI_Info` objects that must be reinterpreted as `Hmpih32` or `Hmpih64`, depending on their size, which makes things a bit complicated for the user. In my opinion, it would be great to be able to directly pass a `MPI.Comm` or a `MPI.Info` object, which are wrapper types defined by MPI.jl.

I started working on a solution that uses [Requires](https://github.com/JuliaPackaging/Requires.jl) to define functions such as `h5p_set_fapl_mpio` only if MPI was loaded before HDF5. This allows to have access to types such as `MPI.Comm`, and define specific variants of MPI-IO functions that accept these types. And thanks to Requires, MPI.jl is not added to the dependency list of HDF5.jl.

Before I submit a PR, I wanted to know if there's interest in this kind of solution.

To be a bit more explicit, this is what part of the PR would look like:
```julia
function h5p_set_fapl_mpio(fapl_id, comm::MPI.Comm, info::MPI.Info)
    h5comm = h5_mpihandle(comm.val)
    h5info = h5_mpihandle(info.val)
    h5p_set_fapl_mpio(fapl_id, h5comm, h5info)
end

h5p_set_fapl_mpio(fapl_id, comm::Hmpih32, info::Hmpih32) =
    h5p_set_fapl_mpio32(fapl_id, comm, info)
h5p_set_fapl_mpio(fapl_id, comm::Hmpih64, info::Hmpih64) =
    h5p_set_fapl_mpio64(fapl_id, comm, info)
```

**Comment by ufdup** [comment]
Yes, this looks nice but maybe we should get the test/mpio.jl running first. As of now it is broken due to 2 things: The first is related to how MPI.jl handles communicators nowadays (C-based), and it's an easy fix. The second (I believe) has to do with the library not being able to release objects properly (check the section on finalizers in https://juliaparallel.github.io/MPI.jl/stable/usage/).

After updating the communicator and the info variables to be in-line with MPI.jl, I still get errors like
~~~
#001: H5I.c line 1300 in H5I_dec_app_ref(): can't decrement ID ref count
major: Object atom`
minor: Unable to decrement reference count
~~~

**Comment by jipolanco** [comment]
That seems to be a different problem, unrelated to MPI. I'll open a separate issue.

**Comment by tcspestana** [comment]
Ok, let's have another thread.
Just to confuse things: if the code has a single read instruction, it exists clean.

**Comment by jipolanco** [comment]
Closed by #622
